lib/wallet/src/Cardano/Wallet/Write/Tx/Balance/TokenBundleSize.hs

```diff
@@ -16,73 +13,57 @@ import Prelude
     ( TxSize (..) )
-import Cardano.Wallet.Shelley.Compatibility
-    ( toCardanoValue )
+import Cardano.Wallet.Shelley.Compatibility.Ledger
+    ( toLedger )
 import Cardano.Wallet.Write.Tx
-    ( PParams, RecentEra, ShelleyLedgerEra, withConstraints )
-import Control.DeepSeq
-    ( NFData )
+    ( PParams, RecentEra, ShelleyLedgerEra, Value, Version, withConstraints )
 import Control.Lens
     ( (^.) )
-import GHC.Generics
-    ( Generic )
-import Numeric.Natural
-    ( Natural )
+import Data.IntCast
+    ( intCastMaybe )
 
-import qualified Cardano.Api.Shelley as Cardano
 import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
-import qualified Data.ByteString as BS
-
--- | The maximum size of a serialized 'TokenBundle'.
--- ('_maxValSize' in the Alonzo ledger)
-newtype TokenBundleMaxSize = TokenBundleMaxSize
-    { unTokenBundleMaxSize :: TxSize }
-    deriving (Eq, Generic, Show)
-
-instance NFData TokenBundleMaxSize
+import qualified Data.ByteString.Lazy as BL
 
 -- | Assesses a token bundle size in relation to the maximum size that can be
 --   included in a transaction output.
 --
 -- See 'W.TokenBundleSizeAssessor' for the expected properties of this function.
 --
-mkTokenBundleSizeAssessor :: TokenBundleMaxSize -> TokenBundleSizeAssessor
-mkTokenBundleSizeAssessor maxSize =
-    TokenBundleSizeAssessor { assessTokenBundleSize }
+mkTokenBundleSizeAssessor
+    :: RecentEra era
+    -> PParams (ShelleyLedgerEra era)
+    -> TokenBundleSizeAssessor
+mkTokenBundleSizeAssessor era pp = TokenBundleSizeAssessor $ \tb ->
+    let
+    in
+        if computeTokenBundleSerializedLengthBytes tb ver > maxValSize
+        then TokenBundleSizeExceedsLimit
+        else TokenBundleSizeWithinLimit
   where
-    assessTokenBundleSize tb
-        | serializedLengthBytes <= maxSize' =
-            TokenBundleSizeWithinLimit
-        | otherwise =
-            TokenBundleSizeExceedsLimit
-      where
-        serializedLengthBytes :: TxSize
-        serializedLengthBytes = computeTokenBundleSerializedLengthBytes tb
+    maxValSize :: TxSize
+    maxValSize = TxSize $ withConstraints era $ pp ^. ppMaxValSizeL
 
-        maxSize' :: TxSize
-        maxSize' = unTokenBundleMaxSize maxSize
+    ver :: Version
+    ver = withConstraints era $ pvMajor $ pp ^. ppProtocolVersionL
 
-computeTokenBundleSerializedLengthBytes :: TokenBundle.TokenBundle -> TxSize
-computeTokenBundleSerializedLengthBytes =
-    TxSize
-        . safeCast
-        . BS.length
-        . serialize' shelleyProtVer
-        . Cardano.toMaryValue
-        . toCardanoValue
+computeTokenBundleSerializedLengthBytes
+    :: TokenBundle.TokenBundle
+    -> Version
+    -> TxSize
+computeTokenBundleSerializedLengthBytes tb ver = serSize (toLedger tb)
   where
-    safeCast :: Int -> Natural
-    safeCast = fromIntegral
+    serSize :: Value StandardCrypto -> TxSize
+    serSize v = maybe err TxSize $ intCastMaybe
+        . BL.length
+        $ serialize ver v
```

Comment:
Is the combination `$ . $` intentional? It seems we could replace it with a more idiomatic `. . $`.

For example:
```suggestion
    serSize v
        = maybe err TxSize
        . intCastMaybe
        . BL.length
        $ serialize ver v
```

Reply:
fixed